smartapps/curb/curb-energy-monitor.src/curb-energy-monitor.groovy

```diff
@@ -0,0 +1,477 @@
+
+            getCurbLocations()
+        }
+        if (state.authToken) {
+            success()
+        } else {
+            fail()
+        }
+    } else {
+        log.error "callback() failed oauthState != state.oauthInitState"
+    }
+}
+
+private removeChildDevices(delete) {
+    delete.each {
+        deleteChildDevice(it.deviceNetworkId)
+    }
+}
+
+def getCurbLocations() {
+    def params = [
+        uri: "http://app.energycurb.com",
+        path: "/api/locations",
+        headers: ["Authorization": "Bearer ${state.authToken}"]
+    ]
+
+    try {
+        httpGet(params) {
+            resp ->
+                def locationNameList = []
+            def locationLookup = []
+            resp.data.each {
+                locationNameList.push(it.name)
+                locationLookup.push(it)
+            }
+            state.locationNames = locationNameList
+            state.locationLookup = locationLookup
+        }
+    } catch (e) {
+        log.error "something went wrong: $e"
+    }
+}
+
+def updateSelectedLocationId() {
+  def location = ""
+    state.locationLookup.each {
+        if (it.name == settings.curbLocation) {
+            location = it.id
+        }
+    }
+    state.location = location
+}
+
+def updateChildDevice(dni, label, value) {
+    def device = getChildDevice(dni)
+    if (device == null) {
+      device = createChildDevice(dni, label)
+    }
+    device.handlePower(value)
+}
+
+def createChildDevice(dni, label) {
+    return addChildDevice("curb", "Curb Power Meter", dni, null, [name: "${dni}", label: "${label}"])
+}
+
+def getDevices() {
+    def params = [
+        uri: "https://app.energycurb.com",
+        path: "/api/latest/${state.location}",
+        headers: ["Authorization": "Bearer ${state.authToken}"],
+        requestContentType: 'application/json'
+    ]
+    asynchttp_v1.get(processDevices, params)
+}
+
+def getAllData() {
+
+    def billingParams = [
+        uri: "https://app.energycurb.com",
+        path: "/api/aggregate/${state.location}/billing/h",
+        headers: ["Authorization": "Bearer ${state.authToken}"],
+        requestContentType: 'application/json'
+    ]
+
+    asynchttp_v1.get(processKwh, billingParams)
+
+    def latestparams = [
+        uri: "https://app.energycurb.com",
+        path: "/api/aggregate/${state.location}/1m/s",
+        headers: ["Authorization": "Bearer ${state.authToken}"],
+        requestContentType: 'application/json'
+    ]
+
+    asynchttp_v1.get(processUsage, latestparams)
+
+}
+
+def processUsage(resp, data) {
+    if (resp.hasError()) {
+        refreshAuthToken()
+        log.error "Usage Response Error: ${resp.getErrorMessage()}"
+        return
+    }
+    def json = resp.json
+    def main = 0.0
+    def production = 0.0
+    def consumption = 0.0
+    if (json) {
+        json.each {
+            if (!it.main && !it.production) {
+              updateChildDevice("${it.id}", it.label, it.avg)
```

Comment:
The response from `/api/aggregate/${state.location}/1m/s` does not include a key of `avg`. This will fail and will send a null event.